README.md
# Amazon Personalize Optimizer Using Amazon Pinpoint Events
Enabling the automated process of using the events generated by their Pinpoint campaigns to rapidly train and optimize their Personalize campaigns.

## Running unit tests for customization
* Clone the repository, then make the desired code changes
* Next, run unit tests to make sure added customization passes the tests
```
cd ./deployment
chmod +x ./run-unit-tests.sh  \n
./run-unit-tests.sh \n
```

## Building distributable for customization
* Configure the bucket name of your target Amazon S3 distribution bucket
```
export DIST_OUTPUT_BUCKET=my-bucket-name # bucket where customized code will reside
export SOLUTION_NAME=my-solution-name
export VERSION=my-version # version number for the customized code
```
_Note:_ You would have to create an S3 bucket with the prefix 'my-bucket-name-<aws_region>'; aws_region is where you are testing the customized solution. Also, the assets in bucket should be publicly accessible.

* Now build the distributable:
```
chmod +x ./build-s3-dist.sh \n
./build-s3-dist.sh $DIST_OUTPUT_BUCKET $SOLUTION_NAME $VERSION \n
```

* Deploy the distributable to an Amazon S3 bucket in your account. _Note:_ you must have the AWS Command Line Interface installed.
```
aws s3 cp ./dist/ s3://my-bucket-name-<aws_region>/$SOLUTION_NAME/$VERSION/ --recursive --acl bucket-owner-full-control --profile aws-cred-profile-name \n
```

* Get the link of the solution template uploaded to your Amazon S3 bucket.
* Deploy the solution to your account by launching a new AWS CloudFormation stack using the link of the solution template in Amazon S3.

***

## File Structure

```
|-deployment/
  |-build-s3-dist.sh             [ shell script for packaging distribution assets ]
  |-run-unit-tests.sh            [ shell script for executing unit tests ]
  |-amazon-personalize-optimizer-using-amazon-pinpoint-events.yaml          [ solution CloudFormation deployment template ]
|-source/
  |-01-cleanup-lambda                     [ micoservice to clean up temp files ]
  |-02-export-pinpoint-endpoints-lambda   [ microservice to initiate an Amazon Pinpoint export ]
  |-03-export-status-lambda               [ microservice to check on the export status ]
  |-04-query-augment-start-lambda         [ microservice to initiate an Amazon Athena query ]
  |-05-query-status-lambda                [ microservice to check on the query status ]
  |-06-dataset-import-lambda              [ microservice to initiate an Amazon Personalize dataset import ]
  |-07-dataset-import-status-lambda       [ microservice to check on the import status ]
  |-08-create-solution-version-lambda     [ microservice to initiate an Amazon Personalize solution version ]
  |-09-create-solution-version-status-lambda     [ microservice to check on the solution version ]
  |-10-update-campaign-lambda             [ microservice to initiate an Amazon Personalize campaign update ]
  |-11-update-campaign-status-lambda      [ microservice to check on the campaign update status ]
  |-custom-resource-helper                [ microservice to help deploy queries during CF deployment ]
  |-firehose-record-conversion-lambda     [ microservice to transform the Amazon Pinpoint event stream events ]
```

Each microservice follows the structure of:

```
|-service-name/
  |-index.js [injection point for microservice]
  |-package.json
```

***


Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.

Licensed under the Apache License Version 2.0 (the "License"). You may not use this file except in compliance with the License. A copy of the License is located at

    http://www.apache.org/licenses/

or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.0.1] - 2020-05-19
### Updated
- Updated the Role permissions for DataSetImportLambdaRole
- Updated Data Lake S3 bucket policy to only allow SSL traffic


## [1.0.0] - 2020-02-10
### Added
- initial repository version


deployment/build-s3-dist.sh
#!/bin/bash
#
# This assumes all of the OS-level configuration has been completed and git repo has already been cloned
#
# This script should be run from the repo's deployment directory
# cd deployment
# ./build-s3-dist.sh source-bucket-base-name solution-name version-code
#
# Paramenters:
#  - source-bucket-base-name: Name for the S3 bucket location where the template will source the Lambda
#    code from. The template will append '-[region_name]' to this bucket name.
#    For example: ./build-s3-dist.sh solutions my-solution v1.0.0
#    The template will then expect the source code to be located in the solutions-[region_name] bucket
#
#  - solution-name: name of the solution for consistency
#
#  - version-code: version of the package

# Check to see if input has been provided:
if [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ]; then
    echo "Please provide the base source bucket name, trademark approved solution name and version where the lambda code will eventually reside."
    echo "For example: ./build-s3-dist.sh solutions trademarked-solution-name v1.0.0"
    exit 1
fi

# Get reference for all important folders
template_dir="$PWD"
template_dist_dir="$template_dir/global-s3-assets"
build_dist_dir="$template_dir/regional-s3-assets"
source_dir="$template_dir/../source"

echo "------------------------------------------------------------------------------"
echo "[Init] Clean old dist, node_modules and bower_components folders"
echo "------------------------------------------------------------------------------"
echo "rm -rf $template_dist_dir"
rm -rf $template_dist_dir
echo "mkdir -p $template_dist_dir"
mkdir -p $template_dist_dir
echo "rm -rf $build_dist_dir"
rm -rf $build_dist_dir
echo "mkdir -p $build_dist_dir"
mkdir -p $build_dist_dir

echo "------------------------------------------------------------------------------"
echo "[Packing] Templates"
echo "------------------------------------------------------------------------------"
echo "cp $template_dir/*.template $template_dist_dir/"
cp $template_dir/*.template $template_dist_dir/
echo "copy yaml templates and rename"
cp $template_dir/*.yaml $template_dist_dir/
cd $template_dist_dir
# Rename all *.yaml to *.template
for f in *.yaml; do
    mv -- "$f" "${f%.yaml}.template"
done

cd ..
echo "Updating code source bucket in template with $1"
replace="s/%%BUCKET_NAME%%/$1/g"
echo "sed -i '' -e $replace $template_dist_dir/*.template"
sed -i '' -e $replace $template_dist_dir/*.template
replace="s/%%SOLUTION_NAME%%/$2/g"
echo "sed -i '' -e $replace $template_dist_dir/*.template"
sed -i '' -e $replace $template_dist_dir/*.template
replace="s/%%VERSION%%/$3/g"
echo "sed -i '' -e $replace $template_dist_dir/*.template"
sed -i '' -e $replace $template_dist_dir/*.template

echo "------------------------------------------------------------------------------"
echo "Build Functions"
echo "------------------------------------------------------------------------------"
echo "Building Function 01-cleanup-lambda"
cd $source_dir/01-cleanup-lambda
npm run build
cp ./dist/01-cleanup-lambda.zip $build_dist_dir/01-cleanup-lambda.zip

echo "Building Function 02-export-pinpoint-endpoints-lambda"
cd $source_dir/02-export-pinpoint-endpoints-lambda
npm run build
cp ./dist/02-export-pinpoint-endpoints-lambda.zip $build_dist_dir/02-export-pinpoint-endpoints-lambda.zip

echo "Building Function 03-export-status-lambda"
cd $source_dir/03-export-status-lambda
npm run build
cp ./dist/03-export-status-lambda.zip $build_dist_dir/03-export-status-lambda.zip

echo "Building Function 04-query-augment-start-lambda"
cd $source_dir/04-query-augment-start-lambda
npm run build
cp ./dist/04-query-augment-start-lambda.zip $build_dist_dir/04-query-augment-start-lambda.zip

echo "Building Function 05-query-status-lambda"
cd $source_dir/05-query-status-lambda
npm run build
cp ./dist/05-query-status-lambda.zip $build_dist_dir/05-query-status-lambda.zip

echo "Building Function 06-dataset-import-lambda"
cd $source_dir/06-dataset-import-lambda
npm run build
cp ./dist/06-dataset-import-lambda.zip $build_dist_dir/06-dataset-import-lambda.zip

echo "Building Function 07-dataset-import-status-lambda"
cd $source_dir/07-dataset-import-status-lambda
npm run build
cp ./dist/07-dataset-import-status-lambda.zip $build_dist_dir/07-dataset-import-status-lambda.zip

echo "Building Function 08-create-solution-version-lambda"
cd $source_dir/08-create-solution-version-lambda
npm run build
cp ./dist/08-create-solution-version-lambda.zip $build_dist_dir/08-create-solution-version-lambda.zip

echo "Building Function 09-create-solution-version-status-lambda"
cd $source_dir/09-create-solution-version-status-lambda
npm run build
cp ./dist/09-create-solution-version-status-lambda.zip $build_dist_dir/09-create-solution-version-status-lambda.zip

echo "Building Function 10-update-campaign-lambda"
cd $source_dir/10-update-campaign-lambda
npm run build
cp ./dist/10-update-campaign-lambda.zip $build_dist_dir/10-update-campaign-lambda.zip

echo "Building Function 11-update-campaign-status-lambda"
cd $source_dir/11-update-campaign-status-lambda
npm run build
cp ./dist/11-update-campaign-status-lambda.zip $build_dist_dir/11-update-campaign-status-lambda.zip

echo "Building Function firehose-record-conversion-lambda"
cd $source_dir/firehose-record-conversion-lambda
npm run build
cp ./dist/firehose-record-conversion-lambda.zip $build_dist_dir/firehose-record-conversion-lambda.zip

echo "Building Function custom-resource-helper"
cd $source_dir/custom-resource-helper
npm run build
cp ./dist/custom-resource-helper.zip $build_dist_dir/custom-resource-helper.zip

# Finish Up
echo "Completed building distribution"
cd $template_dir
